sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -146,21 +160,150 @@ public String getMimeType() {
+        this.numShards = numShards;
+      }
+
+      public int getShardNumber() {
+        return shardNumber;
+      }
+
+
+      public int getNumShards() {
+        return numShards;
+      }
+    }
+
+    /**
+     * Context used for generating a name based on window, pane, shard numer, and num shards.
+     * The policy must produce unique filenames for unique {@link WindowedContext} objects.
+     */
+    public static class WindowedContext extends Context {
+      @Nullable private BoundedWindow window;
+      @Nullable private PaneInfo paneInfo;
+
+      public WindowedContext(
+          BoundedWindow window,
+          PaneInfo paneInfo,
+          int shardNumber,
+          int numShards) {
+        super(shardNumber, numShards);
+        this.window = window;
+        this.paneInfo = paneInfo;
+      }
+
+      @Nullable
+      public BoundedWindow getWindow() {
+        return window;
+      }
+
+      @Nullable
+      public PaneInfo getPaneInfo() {
+        return paneInfo;
+      }
+    }
+
+    /**
+     * When a sink has requested windowed or triggered output, this method will be invoked to return
+     * the filename. The {@link WindowedContext} object gives access to the window and pane, as
+     * well as sharding information. The policy must return unique and consistent filenames
+     * for different windows and panes.
+     */
+    public abstract String windowedFilename(WindowedContext c);
+
+    /**
+     * When a sink has not requested windowed output, this method will be invoked to return the
+     * filename. The {@link Context} object only provides sharding information, which is used by
+     * the policy to generate unique and consistent filenames.
+     */
+    public abstract String unwindowedFilename(Context c);
+
+    /**
+     * @return The base filename for all output files.
+     */
+    public abstract ValueProvider<String> getBaseOutputFilenameProvider();
+
+    /**
+     * Populates the display data.
+     */
+    public void populateDisplayData(DisplayData.Builder builder) {
+    }
+  }
 
   /**
-   * The extension to be used for the final output files.
+   * A default filename policy.
    */
-  protected final String extension;
+  protected class DefaultFilenamePolicy extends FilenamePolicy {
+    ValueProvider<String> baseOutputFilename;
+    String extension;
+    String fileNamingTemplate;
+
+    public DefaultFilenamePolicy(ValueProvider<String> baseOutputFilename, String extension,
+                                 String fileNamingTemplate) {
+      this.baseOutputFilename = baseOutputFilename;
+      if (!isNullOrEmpty(writableByteChannelFactory.getFilenameSuffix())) {
+        this.extension = extension + getFileExtension(
+            writableByteChannelFactory.getFilenameSuffix());
+      } else {
+        this.extension = extension;
+      }
+      this.fileNamingTemplate = fileNamingTemplate;
+    }
+
+    @Override
+    public String unwindowedFilename(FilenamePolicy.Context context) {
```

Comment:
Just inline the few params from the contexts. Otherwise you still have the problem of adding information to the context that are required to be handled and some old policy doesn't notice. We pretty explicitly don't want smooth compatibility here.